app/sidero-controller-manager/cmd/events-manager/adapter.go

```diff
@@ -91,3 +140,77 @@ func (a *Adapter) HandleEvent(ctx context.Context, event events.Event) error {
 
 	return nil
 }
+
+func (a *Adapter) updateAddresses(ctx context.Context, ip string, event *machine.AddressEvent) error {
+	a.nodesMu.Lock()
+	defer a.nodesMu.Unlock()
+
+	name, ok := a.nodes[ip]
+	if !ok {
+		return fmt.Errorf("failed to find ServerBindings for ip %s", ip)
+	}
+
+	var serverbinding sidero.ServerBinding
+	if err := a.metalClient.Get(ctx, name, &serverbinding); err != nil {
+		return err
+	}
+
+	patchHelper, err := patch.NewHelper(&serverbinding, a.metalClient)
+	if err != nil {
+		return err
+	}
+
+	serverbinding.Spec.Addresses = event.Addresses
+	serverbinding.Spec.Hostname = event.Hostname
+
+	return patchHelper.Patch(ctx, &serverbinding)
+}
+
+func (a *Adapter) notify(old, new interface{}) {
+	var oldServerBinding, newServerBinding *sidero.ServerBinding
+
+	if old != nil {
+		oldServerBinding = &sidero.ServerBinding{}
+
+		err := runtime.DefaultUnstructuredConverter.
+			FromUnstructured(old.(*unstructured.Unstructured).UnstructuredContent(), oldServerBinding)
+		if err != nil {
+			a.logger.Error("failed converting old event object", zap.Error(err))
+
+			return
+		}
+	}
+
+	if new != nil {
+		newServerBinding = &sidero.ServerBinding{}
+
+		err := runtime.DefaultUnstructuredConverter.
+			FromUnstructured(new.(*unstructured.Unstructured).UnstructuredContent(), newServerBinding)
+		if err != nil {
+			a.logger.Error("failed converting new event object", zap.Error(err))
+
+			return
+		}
+	}
+
+	a.nodesMu.Lock()
+	defer a.nodesMu.Unlock()
+
+	if new == nil {
+		delete(a.nodes, oldServerBinding.Spec.SideroLink.NodeAddress)
+	} else {
+		address := newServerBinding.Spec.SideroLink.NodeAddress
```

Comment:
if serverBinding got update, we should remove old address the map, to delete unused addresses